### PR TITLE
[SPARK-35988][SS] The implementation for RocksDBStateStoreProvider

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Add the implementation for the RocksDBStateStoreProvider. It's the subclass of StateStoreProvider that leverages all the functionalities implemented in the RocksDB instance.

### Why are the changes needed?
The interface for the end-user to use the RocksDB state store.

### Does this PR introduce _any_ user-facing change?
Yes. New RocksDBStateStore can be used in their applications.

### How was this patch tested?
New UT added.

### CODE REVIEW

**Review by HeartSaVioR — COMMENTED** [review]
Looks OK, but I'd like to make sure @viirya is also OK with the change, as I'm familiar with this change and might be biased.

**Comment by HeartSaVioR** [comment]
NOTE: This might have some post-review comments from #32934 .

**Comment by HeartSaVioR** [comment]
Finally we only have this one for RocksDB state store provider! Please rebase this so that we can continue.

**Comment by xuanyuanking** [comment]
Yes, finally!
Addressed the comments and rebased the code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Thanks, done in 9ddbf58

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:195]: Yes, the dir contains `=()` works in the Linux file system.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:246]: Yea, agree. For the new encoding version, we should have branches here for different versions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:264]: Copy that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:319]: Thanks, done in 9ddbf58

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:110]: nit: I guess we can loose the condition via `> 0L`. "The number of files is 1" doesn't seem to something we need to check.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:239]: Just a head-up: "maintenance" and "snapshotting" are moved from StateStoreSuiteBase to StateStoreSuite as these tests cannot be applied to RocksDB state store provider.

**Review by HeartSaVioR — COMMENTED** [review]
The code change looks good to me. Great!

One thing we haven't been addressed yet is documentation. End users have no idea how to use this provider. Probably we should introduce RocksDB state store provider in SS guide doc.

It's OK if you prefer to file a "blocker" JIRA issue to address documentation separately. I can approve this one then.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala:239]: Yes, since these tests have the assumption that the provider have a `baseDir`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:110]: Agree, will update in the next commit.

**Comment by xuanyuanking** [comment]
```
One thing we haven't been addressed yet is documentation. End users have no idea how to use this provider. Probably we should introduce RocksDB state store provider in SS guide doc.

It's OK if you prefer to file a "blocker" JIRA issue to address documentation separately. I can approve this one then.
```
Sure, create the blocker issue SPARK-36041 for tracking the documentation.

**Review by HeartSaVioR — APPROVED** [review]
+1 

I'll merge once the test passes.

My apologize on not waiting for reviewing on others, we are far behind on progress (compared to the planned date on RC) and this PR is somehow a blocker for session window stuff. The code has been running on production for years so it won't make some problems, but we are open to deal with post-reviews during QA phase.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:47]: I recommend to be consistent with `HDFSBackedStateStore` where it uses `version` and `newVersion`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:64]: Not sure if this log debug is necessary. Especially key/value are unsafe rows.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:102]: "after already committed" -> "after already committed or aborted".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:207]: I guess this is a typo. "the" -> "that".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:252]: nit: I rarely see style like this in Spark.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:306]: rocksdbCommitPauseBgTime? Or FileSyncTime?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:81]: nit: RockDB -> RocksDB

**Review by viirya — APPROVED** [review]
Looks good. Some minor comments. Feel free to address them in follow-up if you prefer to get this in first.

**Comment by HeartSaVioR** [comment]
I'm OK to wait for addressing these comments - I guess we can still merge this in this week, then it's OK for me.

I'm already working on rebasing #33038 on top of this PR. Looks like I need to bring another bunch of lines in #33038 to address RocksDB state store.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:47]: Actually for RocksDB, we keep the name `newVersion`. The major difference here is we have a `loadedVersion` concept, so that's why for the provider side, we have a `latestVersion`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:64]: Make sense, let me delete this debug log.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:207]: Ah yea, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:252]: Make sense, done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:306]: Yea, changed to rocksdbFileSyncTime

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreSuite.scala:81]: Thanks, done in the next commit.

**Review by HeartSaVioR — APPROVED** [review]
+1

Thanks! Merging to master/3.2!

**Comment by HeartSaVioR** [comment]
Thanks @xuanyuanking for contribution! Merged to master/3.2 branches.

**Comment by xuanyuanking** [comment]
Thanks @HeartSaVioR and @viirya for all the help and review!
Finally, all the major implementations of the RocksDB state store have been merged!

**Pull request closed without merging**
